# 4.6 Merging & Exporting Data Exercise and Exporting File as Pickle

# Exercise 4.6 – Merging and Exporting Data
# Instacart Basket Analysis

This notebook executes the following project tasks:
1. Import the cleaned Instacart orders dataset (`orders_checked.csv`)
2. Import the prior order–product dataset (`orders_products_prior.csv`)
3. Merge both datasets using the shared `order_id` key
4. Validate the merge using the `_merge` indicator column
5. Check the resulting dataframe shape to confirm expected row volume
6. Export the combined dataset as a Pickle file for efficient storage (`orders_products_combined.pkl`)

All work performed here directly contributes to building the consolidated Instacart dataset for downstream analysis and segmentation.

In [1]:
#Import libraries
import pandas as pd
import numpy as nb
import os

In [3]:
#02. Setting main project path
#Base path to the Instacart project folder
path = r'/Users/jessduong/Documents/CF/Achievement 4_Python/12-2025 Instacart Basket Analysis'

In [4]:
#Import orders_products_prior (new dataset)
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'))

In [5]:
#Import latest, cleanest orders data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [8]:
#Check the output
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [9]:
#Check the output
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_previous_order  first_order_flag  
0                  8                        0.0              True  
1                  7                       15.0             False  
2                 12                       21.0             False  
3                  7                       29.0             False  
4                 15                       28.0             False

In [12]:
#Check shape of new data set
df_ords_prior.shape

(32434489, 4)

In [14]:
#Check shape of cleaned dataset (checked)
df_ords.shape

(3421083, 8)

In [15]:
#Merging datasets
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [16]:
#Check the output
df_merged_large.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_previous_order  first_order_flag  product_id  \
0                  8                        0.0              True         196   
1                  8                        0.0              True       14084   
2                  8                        0.0              True       12427   
3                  8                        0.0              True       26088   
4                  8                        0.0              True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [18]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [19]:
#Outer merge diagnostic check to verify if true match (note overwriting)
pd.merge(df_ords, df_ords_prior, on = 'order_id', how = 'outer', indicator = True)['_merge'].value_counts()

_merge
both          32434489
left_only       206209
right_only           0
Name: count, dtype: int64

In [23]:
#Export to CV
df_merged_large.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.csv'), index=False)

In [25]:
#Step 2: Export merged file in pickle format as "orders_products_combined.pkl."
df_merged_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

#Export of orders–products combined data
The merged orders dataset (`df_merged_large`) was exported to the Prepared Data folder as:
- `orders_products_combined.csv`
- `orders_products_combined.pkl` (primary working file due to size and faster load time)

The final merged dataset contains 32,434,489 rows, each representing a product within an order.